lib/charms/observability_libs/v0/juju_topology.py

```diff
@@ -0,0 +1,230 @@
+        applied to alert rules topology information ensures that
+        evaluation of alert expressions is restricted to the source
+        (charm) from which the alert rules were obtained.
+
+        Args:
+            model: a string name of the Juju model
+            model_uuid: a globally unique string identifier for the Juju model
+            application: an application name as a string
+            unit: a unit name as a string
+            charm_name: name of charm as a string
+        """
+        self.model = model
+        self.model_uuid = model_uuid
+        self.application = application
+        self.charm_name = charm_name
+        self.unit = unit
+
+    @classmethod
+    def from_charm(cls, charm):
+        """Creates a JujuTopology instance by using the model data available on a charm object.
+
+        Args:
+            charm: a `CharmBase` object for which the `JujuTopology` will be constructed
+        Returns:
+            a `JujuTopology` object.
+        """
+        return cls(
+            model=charm.model.name,
+            model_uuid=charm.model.uuid,
+            application=charm.model.app.name,
+            unit=charm.model.unit.name,
+            charm_name=charm.meta.name,
+        )
+
+    @classmethod
+    def from_dict(cls, data: dict):
+        """Factory method for creating `JujuTopology` children from a dictionary.
+
+        Args:
+            data: a dictionary with five keys providing topology information. The keys are
+                - "model"
+                - "model_uuid"
+                - "application"
+                - "unit"
+                - "charm_name"
+                `unit` and `charm_name` may be empty, but will result in more limited
+                labels. However, this allows us to support payload-only charms.
+
+        Returns:
+            a `JujuTopology` object.
+        """
+        return cls(
+            model=data["model"],
+            model_uuid=data["model_uuid"],
+            application=data["application"],
+            unit=data.get("unit", ""),
+            charm_name=data.get("charm_name", ""),
+        )
+
+    def as_dict(
+        self,
+        *,
+        remapped_keys: Dict[str, str] = {},
+        excluded_keys: List[str] = [],
+        uuid_length: Optional[int] = None,
+    ) -> OrderedDict:
+        """Format the topology information into an ordered dict.
+
+        Keeping the dictionary ordered is important to be able to
+        compare dicts without having to resort to deep comparisons.
+
+        Args:
+            remapped_keys: A dictionary mapping old key names to new key names,
+                which will be substituted when invoked.
+            excluded_keys: A list of key names to exclude from the returned dict.
+            uuid_length: The length to crop the UUID to.
+        """
+        model_uuid = self.model_uuid[: uuid_length - 1] if uuid_length else self.model_uuid
```

Comment:
- Need to account for:
  - uuid_length = int(0)
  - negative values
- The `-1` is not needed.

```suggestion
        uuid_length = uuid_length or len(self.model_uuid)
        model_uuid = self.model_uuid[: uuid_length]
```

Reply:
That is very true. Let's verify the uuid in the constructor and drop the defensive code here.

Reply:
The defensive part here is for the incoming `uuid_length` arg, not the constructor arg.

Reply:
What I meant was that the UUID is cut down to size both here and in `as_identifier`, the only two places where it's referenced.

We could try to check for zero length others in both methods, or we could say  `model_uuid` is a required argument which must actually be a UUID. If it doesn't walk like a UUID and talk like a UUID when you instantiate the object, you don't get a Topology.

This ensures that UUIDs are actually UUIDs and not anything like `zyxwtsrq-woo!` so it means what the attribute says it means, and it also means that we can move these checks (length, negative values, etc) to the constructor where we ensure this. We can also trim it down there so it doesn't need to be handled elsewhere, and it's immediately clear to anyone reading the code exactly why their "UUID" comes out as 8 characters in the label.

Since Python can't reasonably enforce types at runtime (especially with 3.5 compatibility), "type checking" in the constructor with these guards (or more reasonably, trying to pass it to `uuid.UUID` and seeing what we get back) is as good as it gets. If that check passes, we don't need to check here, because we know for sure what it is. If that check fails, they don't get a topology object.

To be defensive at that point, making it a read-only property so it can't be changed would work, but it's Python, so they could always just mutate it anyway even with `__`), so we rely on being good citizens.

Reply:
You can have a look at how I addressed it. I think it takes care of most concerns.